src/Service/WebHook/UpdateLineItemsService.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+namespace OAT\SimpleRoster\Service\WebHook;
+
+use Doctrine\ORM\EntityManagerInterface;
+use OAT\SimpleRoster\Repository\LineItemRepository;
+use OAT\SimpleRoster\WebHook\UpdateLineItemCollection;
+use OAT\SimpleRoster\WebHook\UpdateLineItemDto;
+
+class UpdateLineItemsService
+{
+    /** @var LineItemRepository */
+    private $lineItemRepository;
+
+    /** @var EntityManagerInterface */
+    private $entityManager;
+
+    public function __construct(LineItemRepository $lineItemRepository, EntityManagerInterface $entityManager)
+    {
+        $this->lineItemRepository = $lineItemRepository;
+        $this->entityManager = $entityManager;
+    }
+
+    public function handleUpdates(UpdateLineItemCollection $collection): UpdateLineItemCollection
+    {
+        $knownUpdates = $collection
+            ->filter(
+                function (UpdateLineItemDto $dto): bool {
+                    return $dto->getName() === UpdateLineItemDto::NAME;
+                }
+            )
+            ->setStatus(UpdateLineItemDto::STATUS_ERROR);
+
+        $slugs = $knownUpdates
+            ->map(
+                function (UpdateLineItemDto $dto): string {
+                    return (string)$dto->getSlug();
+                }
+            );
+
+        if (count((array)$slugs) < 1) {
+            return $collection;
+        }
```

Comment:
Here it's visible what is one of my concerns regarding this "collection pipeline" approach. I actually like the idea 👍 , but as you can see here with mapping the collection you just suddenly lost your types. You would expect `string[]`, but you get an iterator instead, forcing you to explicitly case it into an array. If you would encapsulate the mapping logic inside the collection, you could work with less code here and with proper types. What do you think?

Reply:
The types were not lost, it's just a matter of taste. I can change it to an array (string[]) if you want. It was me who add an iterator instead on the UpdateLineItemCollection::map. It was more to use OOP instead of primitive data type.